Add Elemental2DomUtil (utility methods for interacting with the DOM)

karreiro opened the pull request:
Regarding the `Elemental2DomUtil`, it's an Elemental2 approach for the deprecated `DOMUtil` (https://github.com/errai/errai/blob/master/errai-common/src/main/java/org/jboss/errai/common/client/dom/DOMUtil.java).

Regarding the `errai-internal-bom/pom.xml`, the `elemental2-dom` artifact is being added to avoid classes referenced via transitive dependencies. Since we're deprecating Errai elements in favor of Elementa2, we should make them available `errai-internal-bom/pom.xml`.

karreiro commented:
@manstis @mbarkley Could you please take a look? :-)

tiagobento commented on errai-internal-bom/pom.xml:
```diff
@@ -302,6 +306,12 @@
+        <groupId>com.google.elemental2</groupId>
+        <artifactId>elemental2-dom</artifactId>
+        <version>${version.com.google.elemental2}</version>
+      </dependency>
```
I think the place to add this dependency would be in `errai-bom` instead of `errai-internal-bom`. `errai-internal-bom` is supposed to contain the errai modules only. IIRC the elemental2 dependency is already listed on `errai-bom`.

karreiro replied:
Thanks for the comment @tiagobento :-) However, I think that the `errai-bom` and the `errai-internal-bom` have different responsibilities.

Many projects using Errai import the `errai-internal-bom` ([drools-wb](https://github.com/kiegroup/drools-wb/blob/master/pom.xml#L73), [jbpm-wb](https://github.com/kiegroup/jbpm-wb/blob/master/pom.xml#L75), [dashbuilder](https://github.com/dashbuilder/dashbuilder/blob/master/pom.xml#L545), [kie-wb-distribution](https://github.com/kiegroup/kie-wb-distributions/blob/master/pom.xml#L66)).\

That allows the users of the Errai API to declare Errai modules and avoid illegal transitive dependencies ([see](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/pom.xml#L72)).

But now, that the Errai elements are deprecated, sounds fair enough to get the Elemental2 **form** Errai. I mean, I don't think that we should add the Elemental2 version in the `drools-wb` (or even, in `droolsjbpm-build-bootstrap`).

WDYT?

tiagobento replied:
Yeah.. I didn't know about this transitive dependencies check you guys have. I guess in this case it makes sense to have this dependency on the `errai-internal-bom`, but then I'd remove it from the `errai-bom`, given that then it would be duplicated inside Errai. 

I'd wait to hear what other people have to say about this matter, though.

karreiro commented:
Jenkins please retest this.

manstis commented on errai-common/src/main/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtil.java:
```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Provides utility methods for interacting with the DOM.
+ * @author Guilherme Carreiro <ggomes@redhat.com>
```
Our preferred _style_ is to not include ```@author``` tags.

karreiro replied:
I think that Errai adopts a different code style.. Am I missing something?

manstis replied:
Probably I am!

manstis commented on errai-common/src/main/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtil.java:
```diff
@@ -0,0 +1,99 @@
+
+        @Override
+        protected void onAttach() {
+            super.onAttach();
```
I'm not understanding the significance of overriding this method!? 

If it has some significance perhaps a comment in the code would be useful?

karreiro replied:
This is a workaround to access the `onAttach` protected method without JSNI.

[Here](https://github.com/errai/errai/blob/master/errai-common/src/main/java/org/jboss/errai/common/client/dom/DOMUtil.java#L439-L441) the previous version :-)

manstis commented on errai-common/src/test/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtilTest.java:
```diff
@@ -0,0 +1,214 @@
+
+/**
+ * Testing Elemental2DomUtil API.
+ * @author Guilherme Carreiro <ggomes@redhat.com>
```
Our preferred _style_ is to not include ```@author``` tags.

karreiro replied:
I think that Errai adopts a different code style.. Am I missing something?

manstis replied:
Probably I am!

mbarkley commented on errai-common/src/main/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtil.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private static void attach(final Widget child) {
+        final Elemental2Widget widget = Js.uncheckedCast(child);
```
To be honest, this makes me uneasy. As far as I know, `Js.cast` and `Js.uncheckedCast` are meant for casting between native JavaScript types, not GWT/Java types. How have you tested this? Does it work in a production compiled showcase?

My concern here is that in certain compilations (like a production compile with full optimizations) this cast might succeed but the method call to onAttach might fail because the underlying object is not the correct type.

karreiro replied:
@mbarkley Yeah, these casts are a little bit nasty. But, unfortunately, I see them as the solution with least impact that we have to solve this problem right now.

I tested that on a production compiled showcase, and the components work as expected. The following Gif shows a component (Accordion) that depends of the `Elemental2DomUtil`:

[gif - interagindo com accordion] 1

Since the signature of the `attach` method is:
```java
void attach(final Widget child)`
````
...and the `Widget` has a protected method called `onAttach`, I don't think that we're going to see a `child` without the `onAttach` method being called.

Am I missing something?

Thanks.

mbarkley replied:
So I gave this a try in the Errai tutorial, and there is an error calling attach when production compiled. I suspect this screenshot was from one of the webapps that is configured to use draft compile, even when you do a `mvn clean package` (without the "full" profile).

[Here](https://github.com/mbarkley/errai-tutorial/tree/unchecked-cast-repro) is a reproducer in the errai-tutorial. Just run `mvn clean package`, deploy the app to WildFly, open the app, and look in the console for the log message "Something went terribly wrong".

The fix, I think, is to change the attach method back to JSNI. This is an instance where JSNI is the only solution until we are ready to migrate to GWT3 and have a GWT3-specific replacement.

karreiro replied:
@mbarkley I recorded the Gif using a war from `kie-wb-distributions` (master) on WildFly 10.1.0.

But, I'm going to change it to JSNI for now :-) Thanks for the tip.

mbarkley commented:
Jenkins, please retest this.

jomarko commented on errai-common/src/main/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtil.java:
```diff
@@ -0,0 +1,99 @@
+     */
+    public static boolean removeAllElementChildren(final Node node) {
+
+        final boolean hadChildren = node.lastChild != null;
```
Maybe could be more defensive: `final boolean hadChildren = node != null && node.lastChild != null;`, not sure let me know what do you think.

mbarkley replied:
I don't think it's a good idea to silently return if a null is passed in. We could enhance it to assert the parameter is non-null, but it's not a big deal for me since a null would cause an NPE on this line where it would be obvious that node was the culprit.

jomarko replied:
@mbarkley thanks for you comment, it sounds reasonable to me, just to explain my point a little bit more, I wanted to prevent NPE because it is public util method, which doesn't mention this NPE behavior in javadoc, that is why I thought NPE shouldn't be thrown so easy, @karreiro leaving the decision on you, if you prefer variant with NPE, could we have mentioned this in javadoc?

karreiro replied:
Thanks for the comment @jomarko.

However, I agree with @mbarkley, a NullPointerException on this line is clear enough.

I prefer to keep it as is, to keep it simple (or simpler).

mbarkley replied:
FWIW the javadoc looks good to me, because it already states that the node parameter must not be null. Saying that a parameter must not be null makes it fairly obvious what the problem was when you get an NPE.

karreiro commented:
Thank you for the valuable review @mbarkley. Regarding the topics that you raised:
- 2-spaces for indentation: fixed;
- errai-internal-bom: fixed;
- injectable-dom-util: fixed;
- concerns about `Widget.onAttach`: answered (in-line).

I hope to have covered everything that you pointed out.

karreiro commented:
@mbarkley Changes applied.

I tested with the [reproducer](https://github.com/mbarkley/errai-tutorial/tree/unchecked-cast-repro) that you provided, and with `kie-wb-distributions` (master with `-Dfull`). Everything is working as expected.
